Bluetooth: controller: Add initial support for chaining in periodic advertising

andrzej-kaczmarek opened the pull request:
This adds initial support for chaining in periodic advertising, i.e. we can send AUX_SYNC_IND followed by sequence of AUX_CHAIN_IND.

New set of APIs to manage chains instead of single PDUs in advertising set is described in relevant commit. AUX_CHAIN_IND packets are not scheduled as they should be, instead they are just sent one after another at constant offset (T_MAFS) on best effort basis since there's no reservation done for complete train. There's no code that can do fragmentation of data coming from host so I added a temporary commit which will add single chain whenever periodic advertising data is changed on a set.

andrzej-kaczmarek committed:
Bluetooth: controller: Add helper to allocate new pdu_adv

This adds helper to always allocate advertising PDU either from memory
pool or pdu_free queue and does not reuse existing PDU in adv_pdu.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

andrzej-kaczmarek committed:
Bluetooth: controller: Fix PKT_AC_US macro

Advertising channel packets do not have MIC, there's no need to have
extra parameter which always has to be set to 0 anyway.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

andrzej-kaczmarek committed:
Bluetooth: controller: Enable back-to-back chaining for periodic adv

This enables chaining ota for periodic advertising. AUX_CHAIN_IND PDUs
will be sent automatically if AuxPtr is detected in preceding PDU.
AuxPtr offset is always set to achieve minimal required frame spacing,
i.e. 300us (T_mafs). AuxPtr in all PDUs in advertising train are
updated on enqueue since PDU spacing is already known at that time so
we do not need to waste time in LLL.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

andrzej-kaczmarek commented:
@ppryga

ppryga-nordic reviewed (changes requested):
I'd like to see a use of the AUX_CHAIN_IND in a set advertising data procedure. I'm especially interested about handling of data update procedure. 

It would required to create complete chain from scratch. In such situation we have to take extended advertising header from last used chain PDUs and move them to new chain. During that process we have to incorporate new periodic advertising data. Am I correct?

Is it possible to see sample application that uses the functionality?

andrzej-kaczmarek committed:
Bluetooth: controller: Optimize lll_adv_pdu_and_extra_data_alloc

This function is the same as lll_adv_pdu_alloc except it also allocates
extra data at the end - it can just use lll_adv_pdu_alloc call to avoid
extensive c&p.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv.c:
```diff
@@ -99,6 +100,8 @@ static inline bool isr_rx_ci_adva_check(uint8_t tx_addr, uint8_t *addr,
 #define PDU_MEM_COUNT      (PDU_MEM_COUNT_MIN + PDU_MEM_FIFO_COUNT)
 #define PDU_POOL_SIZE      (PDU_MEM_SIZE * PDU_MEM_COUNT)
+/* TODO: make configurable or calculate sane value */
+#define LINK_POOL_SIZE     (16)
```
@andrzej-kaczmarek  As acquire and release both are in the thread context,  memq data structure is not required for chaining. Use simple singly linked list to chain ADV PDU buffers. This will save RAM and reduce the code complexity.

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv.c:
```diff
@@ -107,6 +112,11 @@ static struct {
 /* FIFO to return stale AD data PDU buffers from LLL to thread context */
 static MFIFO_DEFINE(pdu_free, sizeof(void *), PDU_MEM_FIFO_COUNT);
 
+static struct {
+	void *free;
+	memq_link_t pool[LINK_POOL_SIZE];
+} mem_link;
+
```
Do not need this.

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv.h:
```diff
@@ -134,6 +147,24 @@ int lll_adv_data_release(struct lll_adv_pdu *pdu);
 struct pdu_adv *lll_adv_pdu_alloc(struct lll_adv_pdu *pdu, uint8_t *idx);
 void lll_adv_prepare(void *param);
 
+int lll_adv_chain_data_init(struct lll_adv_chain_data *cd);
+int lll_adv_chain_data_reset(struct lll_adv_chain_data *cd);
+/* Allocate new chain from chain data */
+struct lll_adv_chain *lll_adv_chain_alloc(struct lll_adv_chain_data *cd,
+					  uint8_t *idx);
+/* Append new pdu to chain and return it */
+struct pdu_adv *lll_adv_chain_append(struct lll_adv_chain *ch);
+/* Get pdu from chain */
+struct pdu_adv *lll_adv_chain_get_pdu(struct lll_adv_chain *ch,
+				      struct pdu_adv *prev);
+/* Enqueue chain as next one to go in chain data */
+void lll_adv_chain_enqueue(struct lll_adv_chain_data *cd, uint8_t idx);
+/* Get 'last' chain (i.e. most recently enqueued) from chain data */
+struct lll_adv_chain *lll_adv_chain_last_get(struct lll_adv_chain_data *cd);
+/* Get 'current' (i.e. used for TX) chain from chain data */
+struct lll_adv_chain *lll_adv_chain_current_get(struct lll_adv_chain_data *cd,
+						uint8_t *is_modified);
+
```
Use singly linked list. Do not need these implementation.

cvinayak reviewed (changes requested):
Please setup a code walk-through meeting to sync-up on the design.

cvinayak commented:
@wopu-ot Please provide your design thoughts. I had decide during the introduction of 255 byte Advertising PDU to maintain the double buffering design for the advertising PDUs.

For chaining, my proposal is to use singly linked list, wherein:
1. Allocation of first PDU in chaining is using `lll_adv_pdu_alloc` , https://github.com/zephyrproject-rtos/zephyr/blob/3fe33fe3883ffabb8a92bc59debd605e50c26ad7/subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv.c#L256

2. Allocation of subsequent PDUs in chaining is using `adv_pdu_allocate`, https://github.com/zephyrproject-rtos/zephyr/blob/3fe33fe3883ffabb8a92bc59debd605e50c26ad7/subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv.c#L609

Repeated call to  `lll_adv_pdu_alloc`, always returns the first PDU in the chain under preparation. This way HCI commands to append fragments can append PDUs at the tail of the singly linked list.

3. Release of all PDUs in a stale chain by using `lll_adv_pdu_latest_get`, which enqueue each PDU into the MFIFO, https://github.com/zephyrproject-rtos/zephyr/blob/3fe33fe3883ffabb8a92bc59debd605e50c26ad7/subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv.c#L286

andrzej-kaczmarek committed:
Bluetooth: controller: Add support for linked adv PDUs

This adds support to allow advertising PDUs to be linked which is
required to send advertising trains, i.e. AUX_CHAIN_IND.

PDUs are linked with a simple single-linked list, the pointer to next
PDU is stored at the very end of PDU buffer. This prevents it from
being overwritten if PDU is modified and allows for build time offset
calculation.

There are few helpers added to make handling easier, e.g.:
- get next linked PDU
- get last linked PDU
- link one PDU to another
- link PDU at the end
- release all linked PDUs (except for 1st)

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

andrzej-kaczmarek committed:
Bluetooth: controller: Add helper to initialize adv PDU from scratch

This adds adv_sync_pdu_init helper which initializes pdu_adv buffer
with contents of AUX_SYNC_IND/AUX_CHAIN_IND PDU. Extended header flags
can be specified to reserve required space for corresponding fields if
necessary.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

andrzej-kaczmarek committed:
Bluetooth: controller: Add initial support to update AD in chain

This adds some initial support to update AD in chain. We still only
support placing AD in 1st PDU, but this will properly copy any linked
PDUs that may be added due to e.g. CTEInfo present.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

andrzej-kaczmarek committed:
Bluetooth: controller: Add helper to update CTEInfo

This adds simple helper to update CTEInfo. It assumes proper periodic
adv PDU as input to simplify code.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

andrzej-kaczmarek commented:
There are some new helpers to manage PDUs, e.g. duplicate them or create new one based on old one. Not sure if this is the best approach but we need something to conveniently create/duplicate PDUs and link them into chains - I don't think `ull_adv_sync_pdu_set_clear` alone can do everything internally unless it will become really complicated.

This also does not yet allow to accept fragmented data from host, AD is only allowed in 1st PDU as previously but chain can be preserved when updating AD after CTEInfo was added.

andrzej-kaczmarek commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_pdu.h:
```diff
@@ -4,6 +4,16 @@
+#define PDU_ADV_MEM_SIZE       MROUND(PDU_AC_LL_HEADER_SIZE + \
+				      PDU_AC_PAYLOAD_SIZE_MAX + 4)
+#define PDU_ADV_NEXT_PTR(p)    *(struct pdu_adv **)((uint8_t *)(p) + \
+						    PDU_ADV_MEM_SIZE - 4)
```
This places pointer at the very end of PDU buffer which is convenient for two reasons:
- it will not be overwritten when updating PDU contents
- it can be calculated by compiler
The slight disadvantage is that MROUND macro is used here so anything that includes this file needs to include util/mem.h as well.

Another option would be to place pointer just after PDU data - it can be easily calculated at runtime, but may be corrupted when PDU is modified so need to take extra care.

Placing pointer at the beginning of buffer is not a good option since it requires extra handling for PDU data which can start on different offset depending on configuration.

andrzej-kaczmarek replied:
will change to `sizeof(uintptr_t)`

andrzej-kaczmarek commented:
I think I addressed comments that were still relevant after changing to new chaining design as proposed by Vinayak.

andrzej-kaczmarek commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -378,11 +378,27 @@ uint8_t ll_adv_sync_param_set(uint8_t handle, uint16_t interval, uint16_t flags)
+		/* Link to 1st PDU */
+		ter_pdu = (void *)lll_sync->data.pdu[ter_idx];
+		lll_adv_pdu_linked_append(test_pdu, ter_pdu);
+	} while (0);
```
@ppryga This is similar to what (assuming my proposal makes sense) needs to be done when adding CTEInfo: need to allocate new empty PDU, initialize it with CTEInfo header (2nd parameter to adv_sync_pdu_init) and optionally AuxPtr (for all but last PDU), add CTEInfo (adv_sync_cte_info_set) and then link it to parent PDU. Such chain will be preserved if AD is changed, i.e. it will be duplicated as-is.

ppryga replied:
@andrzej-kaczmarek I'd like try to write small test code that will check this approach. I'll try to do it in the beginning of next week.

andrzej-kaczmarek commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv.c:
```diff
@@ -354,11 +354,34 @@ struct pdu_adv *lll_adv_pdu_latest_get(struct lll_adv_pdu *pdu,
+			/* Store partial list in current data index if there is
+			 * no free slot in mfifo. It can be released on next
+			 * switch attempt (on next event).
+			 */
```
Currently there will be no next switch attempt because caller will just assert, as it does without linking. So we can assume this is ok as there always should be enough free slots in mfifo or we can add handling in caller later, i.e. it will skip an event.

Alternatively we can put only 1st PDU in mfifo with all linked PDUs and let ULL deal with it, i.e. it will dequeue this PDU and use it, then release linked PDUs directly to pool.

cvinayak commented on subsys/bluetooth/controller/Kconfig.ll_sw_split:
```diff
@@ -161,6 +161,14 @@ config BT_CTLR_ADV_SYNC_SET
 	help
 	  Maximum supported periodic advertising sets.
 
+config BT_CTLR_ADV_SYNC_PBACK
```
What about `BT_CTLR_ADV_SYNC_PDU_BACK2BACK`?

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.c:
```diff
@@ -583,6 +653,19 @@ void radio_switch_complete_and_tx(uint8_t phy_rx, uint8_t flags_rx, uint8_t phy_
 #endif /* !CONFIG_BT_CTLR_TIFS_HW */
 }
 
+void radio_switch_complete_and_pback(uint8_t phy, uint8_t flags)
```
```suggestion
void radio_switch_complete_and_b2b_tx(uint8_t phy, uint8_t flags)
```

i.e. if using the term back-to-back is ok? That is what is happening in reality. I am ok with piggyback/pback.

andrzej-kaczmarek replied:
back2back sounds better, will use it

cvinayak reviewed (changes requested):
Good work. Minor request to re-use sw_switch, otherwise lets meet for "one last time" to review change requested by @ppryga and settle for a list of pending changes to be reworked.

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -178,11 +178,10 @@ static uint8_t adv_sync_pdu_init_from(struct pdu_adv *pdu,
 	}
 
-	/* Copy AuxPtr, if applicable */
+	/* Add AuxPtr, if applicable. Do not copy since it will be updated later
```
Please squash this commit with previous one. It is unnecessary to alter the same conde one commit after another.

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -271,6 +271,29 @@ static uint8_t adv_sync_pdu_ad_data_set(struct pdu_adv *pdu,
 	return 0;
 }
 
+static uint8_t adv_sync_pdu_cte_info_set(struct pdu_adv *pdu,
```
Please add comment about assumption related with a PDU content - that it is already properly prepared and does not include any field that may follow CTEInfo (AuxPtr, TxPower, ACAD). 

Also I suggest to add assertions to check if AuxPtr, TxPower, ACAD were available, it may save us trouble searching for error if those field were available and overwritten by accident.

andrzej-kaczmarek replied:
This only updates CTEInfo if present, it does not affect other fields in any way so it does not matter if they are there or not.

ppryga replied:
Ok function is named `xxx_set` so the PDU should have correctly allocated memory CTEInfo and PDU is not damaged then. 
Nevertheless please provide some information about that in comment.

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.c:
```diff
@@ -554,6 +554,76 @@ static void sw_switch(uint8_t dir, uint8_t phy_curr, uint8_t flags_curr, uint8_t
 	sw_tifs_toggle &= 1;
 }
+
+static void sw_switch_pback(uint8_t phy, uint8_t flags)
```
Something that I am trying to use for BIS subevents: https://github.com/zephyrproject-rtos/zephyr/pull/33118

andrzej-kaczmarek replied:
`sw_switch` is not inlined so it will not be optimized (unless I misunderstood what you meant) for new case, but regardless reusing `sw_switch` as in mentioned PR works for me since it's just a single extra branch so not much of a difference. I initially thought that there will be more to change there thus I created a copy.

andrzej-kaczmarek commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_sync.c:
```diff
@@ -195,3 +223,142 @@ static int prepare_cb(struct lll_prepare_param *p)
+#endif /* CONFIG_BT_CTLR_GPIO_LNA_PIN */
+
+	if (IS_ENABLED(CONFIG_BT_CTLR_PROFILE_ISR)) {
+		/* NOTE: as scratch packet is used to receive, it is safe to
```
c&p issue :)

andrzej-kaczmarek commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -76,6 +76,51 @@ static void ticker_op_cb(uint32_t status, void *param);
+	LL_ASSERT(!ext_hdr->adi);
+	LL_ASSERT(!ext_hdr->sync_info);
+
+	if (ext_hdr->cte_info) {
```
I can change this but I guess this is just a matter of preference. The way I read this code is that I check flags inside PDU which is initialized earlier using `ext_hdr_flags`, thus I do not check that parameter explicitly since.

ppryga replied:
For me it makes the code more unclear.

andrzej-kaczmarek commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_pdu.h:
```diff
@@ -116,6 +120,9 @@ static inline void lll_adv_sync_data_release(struct lll_adv_sync *lll)
 					     uint8_t idx)
 {
+#if defined(CONFIG_BT_CTLR_ADV_SYNC_PBACK)
+	lll_adv_sync_update_pback(lll, idx);
```
It's convenient to update AuxPtr here because it guarantees that it will be updated every time new data is enqueued and in ULL context we do not need to care about timing (too much). It also does not make any difference for the caller because it still enqueues data as usual and has no side-effects. I can add a comment explaining why this is done here. So I can add short comment to explain why this is done here, but keep function name to be consistent with other calls.

ppryga replied:
In my humble opinion, convenience is not enough to make a function to have hidden side effects. It should be "visible" on the caller site what function actually does. Reader should not be forced to check each function call to see if there is something unexpected done inside.

andrzej-kaczmarek commented:
I included b2b radio interface from https://github.com/zephyrproject-rtos/zephyr/pull/33118 instead of previous implementation, works fine.

ppryga commented:
@andrzej-kaczmarek I've provided a PR to your fork repository that includes some commits that suggest fixes for issues I've found while testing the implementation provided by this PR. (for ref. to others here is link to mentioned PR: https://github.com/andrzej-kaczmarek/zephyr/pull/1).

I've got two additional things:

-  what do you think about refactorization of functions that do operations on periodic advertising PDUs? I mean former code that is already in upstream and your new code. It looks like it may be made more common.
- your design assumes that every time PDUs are updated new chain is created. In my humble opinion new chain should be created only in situation that former one is already in use by LLL (it was enqueued to LLL and peeked by it). In case we update the chain (e.g. we have set new periodic advertising data and want to enable CTE), it seems to be a better option to re-use existing chain:

1. update PDUs is possible
2. add new PDUs to chain if required (e.g. number of CTE is greater than previous length of chain)
3. remove (and release memory) if we find out that there are empty PDUs that do nothing.
 
Thanks to re-use of existing chain we save a lot of memory. To handle CTE (max number of PDUs is 16) we need ~4K of RAM. 
If we want to have memory for: chain in use by LLL, chain that was updated and chain for future updates we need ~12K of RAM.
If we re-use existing chain we save 4K memory.

andrzej-kaczmarek commented:
> @andrzej-kaczmarek I've provided a PR to your fork repository that includes some commits that suggest fixes for issues I've found while testing the implementation provided by this PR. (for ref. to others here is link to mentioned PR: [andrzej-kaczmarek#1](https://github.com/andrzej-kaczmarek/zephyr/pull/1)).

Would be better if we stick to traditional GitHub way and have comments here. It's not quite easy to figure out which parts of that PR are purely fixes and which are added for CTE.

> I've got two additional things:
> 
>     * what do you think about refactorization of functions that do operations on periodic advertising PDUs? I mean former code that is already in upstream and your new code. It looks like it may be made more common.

I think the main question here whether we want to have one big function like `adv_sync_hdr_set_clear` that will just do whatever it takes to update chain with new/removed fields or set of small helpers that can be combined in each case (e.g. set data, enable CTE) to do specific task. I prefer the latter, but don't know what others think about it.

>     * your design assumes that every time PDUs are updated new chain is created. In my humble opinion new chain should be created only in situation that former one is already in use by LLL (it was enqueued to LLL and peeked by it). In case we update the chain (e.g. we have set new periodic advertising data and want to enable CTE), it seems to be a better option to re-use existing chain:
> 
> 
>     1. update PDUs is possible
> 
>     2. add new PDUs to chain if required (e.g. number of CTE is greater than previous length of chain)
> 
>     3. remove (and release memory) if we find out that there are empty PDUs that do nothing.

I can update set data to reuse old PDUs.

> Thanks to re-use of existing chain we save a lot of memory. To handle CTE (max number of PDUs is 16) we need ~4K of RAM.
> If we want to have memory for: chain in use by LLL, chain that was updated and chain for future updates we need ~12K of RAM.
> If we re-use existing chain we save 4K memory.

That could be improved if we have pools of different buffer sizes so small PDUs do not waste memory.

cvinayak commented:
> I think the main question here whether we want to have one big function like adv_sync_hdr_set_clear that will just do whatever it takes to update chain with new/removed fields or set of small helpers that can be combined in each case (e.g. set data, enable CTE) to do specific task. I prefer the latter, but don't know what others think about it.

@andrzej-kaczmarek Please refer to this discussion, and lets align, so we dont have redundancy in our efforts. https://github.com/zephyrproject-rtos/zephyr/pull/32622#discussion_r587425119

andrzej-kaczmarek commented:
> > I think the main question here whether we want to have one big function like adv_sync_hdr_set_clear that will just do whatever it takes to update chain with new/removed fields or set of small helpers that can be combined in each case (e.g. set data, enable CTE) to do specific task. I prefer the latter, but don't know what others think about it.
> 
> @andrzej-kaczmarek Please refer to this discussion, and lets align, so we dont have redundancy in our efforts. [#32622 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/32622#discussion_r587425119)

@cvinayak does it mean that `adv_sync_hdr_set_clear` should be do-it-all, i.e. should also handle chain internally?

cvinayak commented:
@andrzej-kaczmarek I am open to a discussion and settle on a solution based on pros and cons. This will have to be a group meeting after the Easter vacations.

@wopu-ot Could you please include this as a action item that we can settle ASAP (just after Easter vacations) ?

ppryga commented:
> Would be better if we stick to traditional GitHub way and have comments here. It's not quite easy to figure out which parts of that PR are purely fixes and which are added for CTE.

@andrzej-kaczmarek Sorry for that. I thought you could just pick and apply fixes I've done when testing the implementation (also provided you code that could be used to check if everything works). Question -> does it mean that I have to provide comments here? Or will you use provided commits?

BTW. My expectation about implementation is that chaining will also correctly handle PDUs that have CTE.
So, I'd prefer to get an API that will allow me create, modify and destroy a chain of PDUs that has included CTE.

andrzej-kaczmarek commented:
> > Would be better if we stick to traditional GitHub way and have comments here. It's not quite easy to figure out which parts of that PR are purely fixes and which are added for CTE.
> 
> @andrzej-kaczmarek Sorry for that. I thought you could just pick and apply fixes I've done when testing the implementation (also provided you code that could be used to check if everything works). Question -> does it mean that I have to provide comments here? Or will you use provided commits?

[`a6e80bd` (#1)](https://github.com/andrzej-kaczmarek/zephyr/pull/1/commits/a6e80bd26608ba823b2607bf31391d37da2ee38d)
[`90a6d01` (#1)](https://github.com/andrzej-kaczmarek/zephyr/pull/1/commits/90a6d01e78a8b9506dcad0f741cbfa93ac5dbba5)

Is there anything else that fixes issue in this PR?

ppryga commented:
> > > Would be better if we stick to traditional GitHub way and have comments here. It's not quite easy to figure out which parts of that PR are purely fixes and which are added for CTE.

> > @andrzej-kaczmarek Sorry for that. I thought you could just pick and apply fixes I've done when testing the implementation (also provided you code that could be used to check if everything works). Question -> does it mean that I have to provide comments here? Or will you use provided commits?

> a6e80bd (#1)
> 90a6d01 (#1)

> Is there anything else that fixes issue in this PR?

@andrzej-kaczmarek 
Here is a list of commits that in my humble opinion should be added to this PR:
- 68c7dac09af48fcc1890f077b8201feb3f17a051 
- c3594e233a54c14df3abbdf6f74ae90fc1e72395
- 74134db1b30b75ca2c0a69a9fa787b3d2b860b82
- a6e80bd26608ba823b2607bf31391d37da2ee38d
- 90a6d01e78a8b9506dcad0f741cbfa93ac5dbba5
- a057ded850b7f86bd42147ebe08a07501662c56e

ppryga commented:
@andrzej-kaczmarek 
> I think the main question here whether we want to have one big function like adv_sync_hdr_set_clear that will just do whatever it takes to update chain with new/removed fields or set of small helpers that can be combined in each case (e.g. set data, enable CTE) to do specific task. I prefer the latter, but don't know what others think about it.

If I may add one thing here, it is better to stick to one approach than doing things differently at the same time. It would be confusing in future if we stay with different approaches.

andrzej-kaczmarek commented:
Updated to use `ull_adv_sync_pdu_set_clear` to handle ADV_SYNC_IND as in current code. Linked PDUs are only duplicated if latest PDU is currently used by LLL, otherwise they are reused as-is in new PDU. Also fixed few issues and rebased.

There's no B2B radio interface included here since after update to latest version of https://github.com/zephyrproject-rtos/zephyr/pull/33118 it does not rebase easily, so for testing you can try https://github.com/andrzej-kaczmarek/zephyr/tree/bt-sync-pb-full which is exact code from this PR + necessary interfaces to run it.

andrzej-kaczmarek commented:
rebased after https://github.com/zephyrproject-rtos/zephyr/pull/33118 was merged

ppryga-nordic reviewed (changes requested):
Added a couple of minor comments. There are still unresolved comments. Also there are missing following things I've suggested as part of PR I've provided to your private fork:
- missing possibility to configure max length of chain ,
- `PDU_MEM_COUNT_MIN` and `PDU_MEM_FIFO_COUNT` hasn't got additional storage for PDUs for chains,
- missing possibility to remove single linked PDU, only `lll_adv_pdu_linked_release_from_ull` that releases whole chain from its head.
- I assume that change https://github.com/zephyrproject-rtos/zephyr/pull/31875/commits/a22dac22af36d458cfb9b13a9209194e4eae0a8a#diff-3184590f25e7d9697815d4c8e4bc868dd3429067a409add63215421214f60eecR760-R781 is related with decrease of memory required for update of PDUs content. Unfortunately I've got a concern about that change and it seems to me that the memory consumption problem is not yet resolved.

carlescufi commented:
@andrzej-kaczmarek please take a look at the review comments.

andrzej-kaczmarek commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv.c:
```diff
@@ -319,9 +324,26 @@ struct pdu_adv *lll_adv_pdu_alloc_pdu_adv(void)
 }
 
+#if defined(CONFIG_BT_CTLR_ADV_PDU_LINK)
+void lll_adv_pdu_linked_release_from_ull(struct pdu_adv *pdu)
```
`from_ull` means this should be called from ULL only
I can change this name to 'release_all' and/or put a comment in header to explain, I do not really think overly verbose function names are good idea since they are just too long, there has to be some compromise made.

ppryga replied:
@cvinayak what do you think about such (`*_from_ull`) convention? If that goes into master then we will establish a scheme that other contributors may follow.

andrzej-kaczmarek commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -77,6 +77,51 @@ static void ticker_op_cb(uint32_t status, void *param);
+
+	ext_hdr = &com_hdr->ext_hdr;
+	*(uint8_t *)ext_hdr = ext_hdr_flags;
+	dptr = ext_hdr->data;
```
this is just a matter of personal preference, it does not change anything

andrzej-kaczmarek commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -573,6 +757,28 @@ uint8_t ull_adv_sync_pdu_set_clear(struct ll_adv_set *adv,
+	 * with chain as it's already there. In other case we need to duplicate
+	 * chain from current PDU and append it to new PDU.
+	 */
+	if (pdu_prev != pdu_new) {
```
> Added a couple of minor comments. There are still unresolved comments. Also there are missing following things I've suggested as part of PR I've provided to your private fork:
> 
>     * missing possibility to configure max length of chain ,

max length of chain is determined by max length of contents which is implicitly determined by max adv data - that setting is already configurable so I don't think explicit limit on chain length makes sense since it only creates extra problems. you may want to limit numer of PDUs with CTEInfo, but that is not something to resolve in this PR.

>     * `PDU_MEM_COUNT_MIN` and `PDU_MEM_FIFO_COUNT` hasn't got additional storage for PDUs for chains,

those are already calculated using max adv data and max fragment length so reserve extra space for PDUs in chains. again, you may want to reserve extra space for PDUs with CTEInfo, but that is not to be handled in this PR.

>     * missing possibility to remove single linked PDU, only `lll_adv_pdu_linked_release_from_ull` that releases whole chain from its head.

I'll add separate call

andrzej-kaczmarek commented:
> Added a couple of minor comments. There are still unresolved comments. Also there are missing following things I've suggested as part of PR I've provided to your private fork:
> 
>     * missing possibility to configure max length of chain ,

max length of chain is determined by max length of contents which is implicitly determined by max adv data - that setting is already configurable so I don't think explicit limit on chain length makes sense since it only creates extra problems. you may want to limit numer of PDUs with CTEInfo, but that is not something to resolve in this PR.

>     * `PDU_MEM_COUNT_MIN` and `PDU_MEM_FIFO_COUNT` hasn't got additional storage for PDUs for chains,

those are already calculated using max adv data and max fragment length so reserve extra space for PDUs in chains. again, you may want to reserve extra space for PDUs with CTEInfo, but that is not to be handled in this PR.

>     * missing possibility to remove single linked PDU, only `lll_adv_pdu_linked_release_from_ull` that releases whole chain from its head.

I'll add separate call

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv.c:
```diff
@@ -281,7 +281,45 @@ struct pdu_adv *lll_adv_pdu_alloc(struct lll_adv_pdu *pdu, uint8_t *idx)
+	return p;
+}
+
+struct pdu_adv *lll_adv_pdu_alloc_pdu_adv(void)
```
Ok, maybe create an issue in Github to enhance it in near future?

andrzej-kaczmarek commented:
New version changes semantics of `ull_adv_sync_pdu_set_clear` as it will now only update data in advertising PDU so it can be used on any PDU in chain. This means we need new call to handle alloc (i.e. double buffering) so `ull_adv_sync_pdu_alloc` is added. It will also handle `extra_data` as this has to be only done once for each new chain. `...alloc` returns prev and next PDUs as well as index for enqueue.

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -558,13 +732,6 @@ uint8_t ull_adv_sync_pdu_set_clear(struct ll_adv_set *adv,
-
 #if defined(CONFIG_BT_CTLR_DF_ADV_CTE_TX)
 	if (extra_data) {
 		adv_sync_extra_data_set_clear(extra_data_prev, extra_data,
```
In my humble opinion if `adv_sync_hdr_set_clear` is extracted from this function, `adv_sync_extra_data_set_clear` should be extracted also. This will make behavior of `ull_adv_sync_pdu_alloc` to be consistent.

`adv_sync_extra_data_set_clear` could be made part of `ull_adv_sync_pdu_set_clear` or it could be extracted to be a global function.

andrzej-kaczmarek replied:
for me it looks like extra data has to be handled only once per chain, so doing this inside alloc makes sense. making it part of ull_adv_sync_pdu_set_clear does not make sense because it would be modified on each PDU in chain while making it global means you need to call it every time alloc is done.

ppryga replied:
If you make `adv_sync_hrd_ser_clear` global, then you have to return information about extra_data. Based on that information you decide if call to `adv_sync_hrd_ser_clear` is required or not. Eventual instruction execution overhead makes code readability and maintenance easier. In my humble opinion it is a good tradeoff. 
I suggest to change that to make the code consistent in behaviour. In case there is needed such optimisation on single instructions level, then we may change that current version.

ppryga-nordic reviewed (changes requested):
Generally looks ok. There are few minor things to be updated.

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_sync.c:
```diff
@@ -255,3 +287,139 @@ static void isr_done(void *param)
+	}
+
+	radio_gpio_lna_setup();
+	radio_gpio_pa_lna_enable(radio_tmr_tifs_base_get() + ADV_SYNC_PDU_B2B_IFS - 4 -
```
ADV_SYNC_PDU_B2B_IFS -> ADV_SYNC_PDU_B2B_AFS?

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -122,8 +122,193 @@ static uint8_t adv_sync_pdu_init(struct pdu_adv *pdu, uint8_t ext_hdr_flags)
+
+	com_hdr = &pdu->adv_ext_ind;
+
+	dptr = &com_hdr->ext_hdr_adv_data[com_hdr->ext_hdr_len];
```
I've just spotted it. `dptr = &com_hdr->ext_hdr_adv_data[com_hdr->ext_hdr_len];` it will overwrite `ext_hdr` (flags octet) in if ext_hdr_len is zero. That will corrupt PDU content with e.g. wrong information about flags.

It will be better to use here `dptr = ext_hdr->data;` to get access to actual PDU data address.

andrzej-kaczmarek replied:
there are no flags if ext_hdr_len==0 because flags are part of ext header

ppryga replied:
ok

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_df.c:
```diff
@@ -295,13 +296,22 @@ uint8_t ll_df_set_cl_cte_tx_enable(uint8_t adv_handle, uint8_t cte_enable)
 			return err;
 		}
 
+		lll_adv_sync_data_enqueue(lll_sync, ter_idx);
```
`lll_adv_sync_data_enqueue` is called once, just after end of if-else statement. It covers both parts of condition. This call is not needed here.

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_df.c:
```diff
@@ -326,13 +336,23 @@ uint8_t ll_df_set_cl_cte_tx_enable(uint8_t adv_handle, uint8_t cte_enable)
 			return err;
 		}
 
+		lll_adv_sync_data_enqueue(lll_sync, ter_idx);
```
`lll_adv_sync_data_enqueue` is called once, just after end of if-else statement. It covers both parts of condition. This call is not needed here.

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -77,8 +69,261 @@ static void ticker_op_cb(uint32_t status, void *param);
+	return 0;
+}
+
+static struct pdu_adv *adv_sync_pdu_duplicate_chain(struct pdu_adv *pdu)
```
It generates: warning: 'adv_sync_pdu_duplicate_chain' defined but not used [-Wunused-function].
Will not pass CI (at least it will make direction finding unit tests to fail on build stage).

andrzej-kaczmarek replied:
hmmm, seems like some piece of code that uses it was deleted by mistake during rebase, will add it again.

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -558,23 +743,99 @@ uint8_t ull_adv_sync_pdu_set_clear(struct ll_adv_set *adv,
+{
+	int err;
+
+	err = adv_sync_hdr_set_clear(lll_sync, ter_pdu_prev, ter_pdu,
```
It looks that `adv_sync_hdr_set_clear` is not needed anymore. Signature of the function `ull_adv_sync_pdu_set_clear` may be changed to use `void* *data` instead of `struct adv_pdu_field_data *data`.

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -558,23 +743,99 @@ uint8_t ull_adv_sync_pdu_set_clear(struct ll_adv_set *adv,
+	 */
+	if (hdr_add_fields & ULL_ADV_PDU_HDR_FIELD_CTE_INFO) {
+		memcpy(extra_data_new, data, sizeof(struct lll_df_adv_cfg));
+	} else if ((hdr_rem_fields & ULL_ADV_PDU_HDR_FIELD_CTE_INFO) ||
```
```suggestion
} else if (!(hdr_rem_fields & ULL_ADV_PDU_HDR_FIELD_CTE_INFO) ||
```
This is an error that is available from first implementation of the code.

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_df.c:
```diff
@@ -295,9 +297,25 @@ uint8_t ll_df_set_cl_cte_tx_enable(uint8_t adv_handle, uint8_t cte_enable)
+		if (extra_data) {
+			ull_adv_sync_extra_data_set_clear(extra_data_prev,
+							  extra_data,
+							  ULL_ADV_PDU_HDR_FIELD_CTE_INFO,
+							  0, NULL);
```
```suggestion
							  0,
							  ULL_ADV_PDU_HDR_FIELD_CTE_INFO, NULL);
```

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -77,6 +77,51 @@ static void ticker_op_cb(uint32_t status, void *param);
 static struct ll_adv_sync_set ll_adv_sync_pool[CONFIG_BT_CTLR_ADV_SYNC_SET];
 static void *adv_sync_free;
 
+static uint8_t adv_sync_pdu_init(struct pdu_adv *pdu, uint8_t ext_hdr_flags)
```
```suggestion
static uint8_t adv_sync_pdu_init(struct pdu_adv *pdu, uint8_t ext_hdr_flags)
```
It always returns zero.

andrzej-kaczmarek committed:
Bluetooth: controller: Remove adv_sync_hdr_set_clear

adv_sync_hdr_set_clear was just wrapped by ull_adv_sync_pdu_set_clear
so we can merge both into single function.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

cvinayak reviewed (approved):
Quick ack, do not have any strong objections, overall design was previously agreed. Any rework if any would be minor, can be taken up in subsequent PRs.

ppryga commented:
Removed part of code responsible for setting address timeout, it should not be configured in `isr_tx` for back to back TX.

cvinayak commented:
@wopu-ot This PR has changed the way extended advertising PDU is allocated, this is conflicting my BIS work, and wish to have the changes here upstreamed before I continue with rework on my PRs. Please do the needful to review this PR.

cvinayak commented:
@ppryga I have rebased to upstream/main and pushed.

ppryga commented:
@nashif, @galak, @carlescufi could one of you merge the PR? CI is green.
